### PR TITLE
sql: add JSON builtins json_typeof

### PR DESCRIPTION
**Pull request by heyihong**
Hi there, I implement the json_typeof function of JSON builtins #20120 in the sql module. Please take a look.

### CODE REVIEW

**Comment by justinj** [comment]
Hi @heyihong! Thanks for much for the contribution! This looks great - just a couple comments.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/json_builtins, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRuHUkIgeThfVpL0E5:-KzRuHUljMXe9FWb4OoJ:b-y4tfoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/logictest/testdata/logic_test/json_builtins#L1)):*
> ```
> # LogicTest: default distsql parallel-stmts
> ```

Good call to make a new file for this, I would have stuffed it in the `json` test file but I think this is cleaner!

---

*[pkg/sql/sem/builtins/builtins.go, line 1499 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRrrrkraD3efB1MNhM:-KzRrrrkraD3efB1MNhN:b89s5v8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/sem/builtins/builtins.go#L1499)):*
> ```Go
> 	},
> 
> 	"json_typeof": {
> ```

We should actually support both `json_typeof` and `jsonb_typeof`, though they should be the same. You can accomplish this by pulling the `tree.Builtin` struct below up into a value and then reusing it across both cases. (CockroachDB only implements what Postgres calls JSONB, but then aliases `JSON` to it).

---

*[pkg/sql/sem/builtins/builtins.go, line 1501 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRsQT1lfPjuHVdfVK9:-KzRsQT1lfPjuHVdfVKA:b8rnqwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/sem/builtins/builtins.go#L1501)):*
> ```Go
> 	"json_typeof": {
> 		tree.Builtin{
> 			Types:      tree.ArgTypes{{"val", types.String}},
> ```

This should operate on `types.JSON`, rather than `types.String`. This appears to work as it does in Postgres because the string literal passed in Postgres is automatically parsed and converted to a JSON datum, but the function itself operates on `JSON`:
```
justin=# select json_typeof('1'::json);
 json_typeof
-------------
 number
(1 row)

justin=# select json_typeof('1'::text);
ERROR:  function json_typeof(text) does not exist
LINE 1: select json_typeof('1'::text);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

---

*[pkg/sql/sem/builtins/builtins.go, line 1504 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRsqbgCtp7kZynKFYm:-KzRsqbgCtp7kZynKFYn:b-oc38v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/sem/builtins/builtins.go#L1504)):*
> ```Go
> 			ReturnType: tree.FixedReturnType(types.String),
> 			Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
> 				str := string(tree.MustBeDString(args[0]))
> ```

That means that the value coming in here will already be a `DJSON`, so there's no need to parse it.

---

*[pkg/util/json/json.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRsyXr2aFcJKlFVjnz:-KzRsyXr2aFcJKlFVjo-:b-413dm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/util/json/json.go#L82)):*
> ```Go
> 
> 	// TypeAsText returns the type of JSON document as a string.
> 	TypeAsText() *string
> ```

This can return a `string`, rather than a `*string`. The reason `AsText` returns `*string` is that it has the possibility of returning `nil` to signal a sql `NULL`, but `TypeAsText` doesn't have that problem.

---

*[pkg/util/json/json.go, line 673 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRtCWYVa3u4TnJgSZF:-KzRtCWYVa3u4TnJgSZG:bzgw93h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/util/json/json.go#L673)):*
> ```Go
> func (j jsonString) TypeAsText() *string {
> 	s := j.jsonType().String()
> 	return &s
> ```

I would just inline the return of the value here, no need to delegate to the type, so
```
func (jsonString) TypeAsText() string { return "string" }
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-:-KzRrhfPFKkOmLLsR8tg:b-74xh97)*

**Comment by heyihong** [comment]
Hi @justinj, Thanks for your careful review. Please take an another look.

---

Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-:-KzSKcAHw8a1F-R4g7m2:bjswb0m)*

File: pkg/sql/sem/tree/datum.go [hunk 1]
```diff
@@ -2074,6 +2074,26 @@ func MakeDJSON(d interface{}) (Datum, error) {
 	return &DJSON{j}, nil
 }
 
+func AsDJSON(e Expr) (*DJSON, bool) {
```
Comment (justinj) [inline pkg/sql/sem/tree/datum.go]: Our CI/linter is complaining because this is an exported function without a comment.

**Comment by justinj** [comment]
All the changes look great! Three final things:

1. Our linter caught a small style nit that needs to be fixed before CI will pass, after fixing that
2. our docs generator needs to be re-run manually on any new builtins, so can you run `make generate PKG=./docs/...`, commit the result and then
3. squash all your commits into a single commit?

Thanks!

**Comment by heyihong** [comment]
Review status: 0 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/json_builtins, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRuHUkIgeThfVpL0E5:-KzURM2pVuSji09m4MCB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/logictest/testdata/logic_test/json_builtins#L1)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Good call to make a new file for this, I would have stuffed it in the `json` test file but I think this is cleaner!
</blockquote></details>

Done.

---

*[pkg/sql/sem/builtins/builtins.go, line 1499 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRrrrkraD3efB1MNhM:-KzURJW_NthMDwD5ar4T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/sem/builtins/builtins.go#L1499)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

We should actually support both `json_typeof` and `jsonb_typeof`, though they should be the same. You can accomplish this by pulling the `tree.Builtin` struct below up into a value and then reusing it across both cases. (CockroachDB only implements what Postgres calls JSONB, but then aliases `JSON` to it).
</blockquote></details>

Done.

---

*[pkg/sql/sem/builtins/builtins.go, line 1501 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRsQT1lfPjuHVdfVK9:-KzURKLtR8H_pDI0fF5v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/sem/builtins/builtins.go#L1501)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This should operate on `types.JSON`, rather than `types.String`. This appears to work as it does in Postgres because the string literal passed in Postgres is automatically parsed and converted to a JSON datum, but the function itself operates on `JSON`:
```
justin=# select json_typeof('1'::json);
 json_typeof
-------------
 number
(1 row)

justin=# select json_typeof('1'::text);
ERROR:  function json_typeof(text) does not exist
LINE 1: select json_typeof('1'::text);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```
</blockquote></details>

Done.

---

*[pkg/sql/sem/builtins/builtins.go, line 1504 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRsqbgCtp7kZynKFYm:-KzURKbieDHCsDHiGjFL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/sql/sem/builtins/builtins.go#L1504)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

That means that the value coming in here will already be a `DJSON`, so there's no need to parse it.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/datum.go, line 2077 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzSJYmLDgZNvSxmsftO-r2-2077:-KzURHW4-tbqddQTtX9I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7570d8e7491177d733bb7223c58e7e86af0c935d/pkg/sql/sem/tree/datum.go#L2077)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Our CI/linter is complaining because this is an exported function without a comment.
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 82 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRsyXr2aFcJKlFVjnz:-KzURI4nrB-oKAaKJTAa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/util/json/json.go#L82)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This can return a `string`, rather than a `*string`. The reason `AsText` returns `*string` is that it has the possibility of returning `nil` to signal a sql `NULL`, but `TypeAsText` doesn't have that problem.
</blockquote></details>

Done.

---

*[pkg/util/json/json.go, line 673 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190#-KzRtCWYVa3u4TnJgSZF:-KzURIPNTmUdI4KFD_Ty:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b2a98aaf4aac5c41d9298ad16e22a92f9562661/pkg/util/json/json.go#L673)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I would just inline the return of the value here, no need to delegate to the type, so
```
func (jsonString) TypeAsText() string { return "string" }
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20190)*

**Comment by justinj** [comment]
Wonderful, thanks for the help. Hope we see you again!

**Pull request merged**